{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Anna Urbala\n",
    "# Implementacja NEAT\n",
    "\n",
    "Z racji tego, że neat jest skomplikowanym podejściem, jego implementacja rozbita jest na dwa tygodnie. W ramach pierwszego deadline należy przesłać niekoniecznie kompletny i działający kod, który będzie dowodem rozpoczęcia pracy nad rozwiązaniem. Dopiero kod w kolejnym tygodniu musi w pełni działać.\n",
    "\n",
    "W tym tygodniu trzeba zaimplementować: kodowanie, w szczególności generowanie fenotypu z genotypu, wyliczanie funkcji dopasowania i część operatora mutacji polegająca na dodaniu neuronu i zmianie wagi.\n",
    "\n",
    "W ramach testu przeprowadzić próbne uczenie (bez krzyżowania) na zbiorze easy z laboratorium o sieciach neuronowych.\n",
    "\n",
    "\n",
    "https://github.com/SirBob01/NEAT-Python/blob/master/neat/neat.py\n",
    "\n",
    "https://github.com/goktug97/NEAT/blob/master/neat/neat.py\n",
    "\n",
    "http://nn.cs.utexas.edu/downloads/papers/stanley.cec02.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# todo\n",
    "# - mutacja biasu\n",
    "# - mutacja funkcji aktywacji\n",
    "from enum import Enum, auto\n",
    "import random\n",
    "from copy import deepcopy\n",
    "import math\n",
    "import operator\n",
    "\n",
    "\n",
    "def sigmoid(x):\n",
    "    return 1 / (1 + math.exp(-x))\n",
    "\n",
    "def relu(x):\n",
    "    return max(x, 0)\n",
    "\n",
    "def tanh(x):\n",
    "    return math.tanh(x)\n",
    "\n",
    "functions = [sigmoid, relu, tanh]\n",
    "\n",
    "class NodeType(Enum):\n",
    "    INPUT = auto()\n",
    "    HIDDEN = auto()\n",
    "    OUTPUT = auto()\n",
    "\n",
    "class ConnectionGene:\n",
    "    def __init__(self, inNode: int, outNode: int, weight=1.0, expressed=True, innovation=0):\n",
    "        self.inNode = inNode\n",
    "        self.outNode = outNode\n",
    "        self.weight = weight\n",
    "        self.expressed = expressed\n",
    "        self.innovation = innovation\n",
    "    \n",
    "    def disable(self):\n",
    "        self.expressed = False\n",
    "\n",
    "class NodeGene:\n",
    "    def __init__(self, type_: NodeType, id_: int, bias: float, fun):\n",
    "        self.type = type_\n",
    "        self.id = id_\n",
    "        self.bias = bias\n",
    "        self.fun = fun\n",
    "\n",
    "class Counter:\n",
    "    def __init__(self):\n",
    "        self.innovation = 0\n",
    "        \n",
    "    def getInnovation(self):\n",
    "        self.innovation += 1\n",
    "        return self.innovation\n",
    "    \n",
    "class Genome:\n",
    "    PROB_PERTURBING = 0.9\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.connections = {}\n",
    "        self.nodes = {}\n",
    "    \n",
    "    def addNodeGene(self, gene: NodeGene):\n",
    "        self.nodes[gene.id] = gene\n",
    "    \n",
    "    def addConnectionGene(self, gene: ConnectionGene):\n",
    "        self.connections[gene.innovation] = gene\n",
    "        \n",
    "    def mutation(self):\n",
    "        for conn in self.connections.values():\n",
    "            if random.uniform(0, 1.0) < self.PROB_PERTURBING:\n",
    "                conn.weight *= random.uniform(-2.0, 2.0)\n",
    "            else:\n",
    "                conn.weight = random.uniform(-2.0, 2.0)\n",
    "            \n",
    "    \n",
    "    def addConnectionMutation(self, innovation: Counter, maxAttempts):\n",
    "        tries = 0\n",
    "        success = False\n",
    "        while tries < maxAttempts and success == False:\n",
    "            tries += 1\n",
    "            node1 = self.nodes[random.randrange(len(self.nodes))]\n",
    "            node2 = self.nodes[random.randrange(len(self.nodes))]\n",
    "\n",
    "            rev = False\n",
    "            if node1.type.value > node2.type.value:\n",
    "                rev = True\n",
    "                node1,node2 = node2,node1\n",
    "            \n",
    "            connectionImpossible = False\n",
    "            if (node1.type == NodeType.INPUT and node2.type == NodeType.INPUT) \\\n",
    "            or (node1.type == NodeType.OUTPUT and node2.type == NodeType.OUTPUT):\n",
    "                connectionImpossible = True\n",
    "\n",
    "            connectionExists = False\n",
    "            for conn in self.connections.values():\n",
    "                if (conn.inNode == node1.id and conn.outNode == node2.id) \\\n",
    "                or (conn.inNode == node2.id and conn.outNode == node1.id):\n",
    "                    connectionExists = True\n",
    "                    break\n",
    "            \n",
    "            if connectionExists or connectionImpossible:\n",
    "                continue\n",
    "\n",
    "            newConn = ConnectionGene(node1.id, node2.id, weight=random.uniform(-1.0, 1.0), innovation=innovation.getInnovation())\n",
    "            self.addConnectionGene(newConn)\n",
    "            success = True\n",
    "    \n",
    "    def addNodeMutation(self, connectionInnovation: Counter, nodeInnovation: Counter):\n",
    "        conn = self.connections[random.randrange(len(self.connections))]\n",
    "        \n",
    "        inNode = self.nodes[conn.inNode]\n",
    "        outNode = self.nodes[conn.outNode]\n",
    "        \n",
    "        conn.disable()\n",
    "        \n",
    "        newNode = NodeGene(NodeType.HIDDEN, nodeInnovation.getInnovation(), random.uniform(-1, 1), random.choice(functions))\n",
    "        newConn1 = ConnectionGene(inNode.id, newNode.id, innovation=connectionInnovation.getInnovation())\n",
    "        newConn2 = ConnectionGene(newNode.id, outNode.id, weight=conn.weight, innovation=connectionInnovation.getInnovation())\n",
    "        \n",
    "        self.addNodeGene(newNode)\n",
    "        self.addConnectionGene(newConn1)\n",
    "        self.addConnectionGene(newConn2)\n",
    "    \n",
    "    \n",
    "    @staticmethod\n",
    "    def crossover(parent1, parent2):\n",
    "        child = Genome()\n",
    "        for parent1Node in parent1.nodes.values():\n",
    "            child.nodes.append(deepcopy(parent1Node))\n",
    "        \n",
    "        for parent1Conn in paren1.connections.values():\n",
    "            childConnGene = deepcopy(parent1Conn)\n",
    "            if parent1Conn.innovation in parent2.connections and random.choice([True, False]):\n",
    "                childConnGene = deepcopy(parent2.connections[parent1Conn.innovation])\n",
    "            child.addConnectionGene(childConnGene)\n",
    "            \n",
    "        return child\n",
    "    \n",
    "    @staticmethod\n",
    "    def compatibilityDistance(genome1, genome2, c1: int, c2: int, c3: int):\n",
    "        excessGenes = Genome.countExcessGenes(genome1, genome2)\n",
    "        disjointGenes = Genome.countDisjointGenes(genome1, genome2)\n",
    "        avgWeightDiff = Genome.averageWeightDiff(genome1, genome2)\n",
    "        return excessGenes * c1 + disjointGenes * c2 + avgWeightDiff * c3\n",
    "    \n",
    "    @staticmethod\n",
    "    def countMatchingGenes(genome1, genome2):\n",
    "        matchingGenes = 0\n",
    "        \n",
    "        highestInnovation1 = max(genome1.nodes.keys())\n",
    "        highestInnovation2 = max(genome2.nodes.keys())\n",
    "        indices = max(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices + 1):\n",
    "            if i in genome1.nodes and i in genome2.nodes:\n",
    "                matchingGenes += 1\n",
    "        \n",
    "        highestInnovation1 = max(genome1.connections.keys())\n",
    "        highestInnovation2 = max(genome2.connections.keys())\n",
    "        indices = max(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices + 1):\n",
    "            if i in genome1.connections and i in genome2.connections:\n",
    "                matchingGenes += 1\n",
    "        \n",
    "        return matchingGenes\n",
    "    \n",
    "    @staticmethod\n",
    "    def countDisjointGenes(genome1, genome2):\n",
    "        disjointGenes = 0\n",
    "        \n",
    "        highestInnovation1 = max(genome1.nodes.keys())\n",
    "        highestInnovation2 = max(genome2.nodes.keys())\n",
    "        indices = min(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices + 1):\n",
    "            if (i not in genome1.nodes and i in genome2.nodes) \\\n",
    "            or (i not in genome2.nodes and i in genome1.nodes):\n",
    "                disjointGenes += 1\n",
    "        \n",
    "        highestInnovation1 = max(genome1.connections.keys())\n",
    "        highestInnovation2 = max(genome2.connections.keys())\n",
    "        indices = min(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices + 1):\n",
    "            if (i not in genome1.connections and i in genome2.connections) \\\n",
    "            or (i not in genome2.connections and i in genome1.connections):\n",
    "                disjointGenes += 1\n",
    "        \n",
    "        return disjointGenes\n",
    "    \n",
    "    @staticmethod\n",
    "    def countExcessGenes(genome1, genome2):\n",
    "        excessGenes = 0\n",
    "        \n",
    "        highestInnovation1 = max(genome1.nodes.keys())\n",
    "        highestInnovation2 = max(genome2.nodes.keys())\n",
    "        indices1 = min(highestInnovation1, highestInnovation2)\n",
    "        indices2 = max(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices1+1, indices2+1):\n",
    "            if (i not in genome1.nodes and i in genome2.nodes) \\\n",
    "            or (i not in genome2.nodes and i in genome1.nodes):\n",
    "                excessGenes += 1\n",
    "        \n",
    "        highestInnovation1 = max(genome1.connections.keys())\n",
    "        highestInnovation2 = max(genome2.connections.keys())\n",
    "        indices1 = min(highestInnovation1, highestInnovation2)\n",
    "        indices2 = max(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices1+1, indices2+1):\n",
    "            if (i not in genome1.connections and i in genome2.connections) \\\n",
    "            or (i not in genome2.connections and i in genome1.connections):\n",
    "                excessGenes += 1\n",
    "        \n",
    "        return excessGenes\n",
    "    \n",
    "    @staticmethod\n",
    "    def averageWeightDiff(genome1, genome2):\n",
    "        matchingGenes = 0\n",
    "        weightDifference = 0.0\n",
    "\n",
    "        highestInnovation1 = max(genome1.connections.keys())\n",
    "        highestInnovation2 = max(genome2.connections.keys())\n",
    "        indices = max(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices + 1):\n",
    "            if i in genome1.connections and i in genome2.connections:\n",
    "                matchingGenes += 1\n",
    "                weightDifference += abs(genome1.connections[i].weight - genome2.connections[i].weight)\n",
    "        \n",
    "        return weightDifference/matchingGenes\n",
    "    \n",
    "    def forward(self, inputs):\n",
    "        \"\"\"Evaluate inputs and calculate the outputs of the\n",
    "        neural network via the forward propagation algorithm.\n",
    "        \"\"\"\n",
    "        _inputs = [node for self.nodes.values() if node.type == NodeType.INPUT]\n",
    "        _inputs.sort(key=operator.attrgetter('id'))\n",
    "\n",
    "        # Set input values\n",
    "        for i, node in enumerate(_inputs):\n",
    "            node.output = inputs[i]\n",
    "        \n",
    "        # Generate backward-adjacency list \n",
    "        _from = {}\n",
    "        _to = {}\n",
    "        waiting_inputs = {}\n",
    "        for node in self.nodes.values():\n",
    "            _from[node.id] = []\n",
    "            _to[node.id] = []\n",
    "            waiting_inputs[node.id] = 0\n",
    "\n",
    "        for conn in self.connections:\n",
    "            if conn.expressed:\n",
    "                _from[conn.outNode].append(conn.inNode)\n",
    "                _to[conn.inNode].append(conn.outNode)\n",
    "                waiting_inputs[conn.outNode] += 1\n",
    "                \n",
    "        ready_nodes = [node for self.nodes.values if waiting_inputs[node.id] == 0]\n",
    "        # todo\n",
    "        while len(ready_nodes):\n",
    "            node = ready_nodes.pop()\n",
    "        for j in ordered_nodes:\n",
    "            ax = 0\n",
    "            for i in _from[j]:\n",
    "                ax += self._edges[(i, j)].weight * self._nodes[i].output\n",
    "\n",
    "            node = self._nodes[j]\n",
    "            node.output = node.activation(ax + node.bias)\n",
    "        \n",
    "        return [self._nodes[n].output for n in range(self._inputs, self._unhidden)]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from abc import ABC, abstractmethod\n",
    "\n",
    "class FitnessGenome:\n",
    "    def __init__(self, genome: Genome, fitness: float):\n",
    "        self.genome = genome\n",
    "        self.fitness = fitness\n",
    "    \n",
    "class Species:\n",
    "    def __init__(self, mascot: Genome):\n",
    "        self.mascot = mascot\n",
    "        self.members = [mascot]\n",
    "        self.fitnessPop = []\n",
    "        self.totalAdjustedFitness = 0.0\n",
    "    \n",
    "    def addAdjustedFitness(self, adjustedFitness: float):\n",
    "        self.totalAdjustedFitness += adjustedFitness\n",
    "    \n",
    "    def reset(self):\n",
    "        newMascotIndex = random.randrange(len(self.members))\n",
    "        self.mascot = self.members[newMascotIndex]\n",
    "        self.members = []\n",
    "        self.fitnessPop = []\n",
    "        self.totalAdjustedFitness = 0.0\n",
    "    \n",
    "class Evaluator(ABC):\n",
    "    C1 = 1.0\n",
    "    C2 = 1.0\n",
    "    C3 = 0.4\n",
    "    DT = 10.0\n",
    "    MUTATION_RATE = 0.8\n",
    "    ADD_CONNECTION_RATE = 0.1\n",
    "    ADD_NODE_RATE = 0.1\n",
    "    CROSSOVER_RATE = 0.75\n",
    "    \n",
    "    def __init__(self, populationSize: int, startingGenome: Genome, nodeInnovation: Counter, connectionInnovation: Counter):\n",
    "        self.speciesMap = {} # genome -> species\n",
    "        self.scoreMap = {} # genome -> float\n",
    "        self.species = []\n",
    "        # self.genomes = []\n",
    "        self.nextGenGenomes = [None for i in range(populationSize)]\n",
    "        self.populationSize = populationSize\n",
    "        self.nodeInnovation = nodeInnovation\n",
    "        self.connectionInnovation = connectionInnovation\n",
    "        self.highestScore = 0.0\n",
    "        self.fittestGenome = None\n",
    "        self.genomes = [deepcopy(startingGenome) for i in range(populationSize)]\n",
    "    \n",
    "    def evaluate():\n",
    "        for s in self.species:\n",
    "            s.reset()\n",
    "        \n",
    "        self.scoreMap = {}\n",
    "        self.speciesMap = {}\n",
    "        self.nextGenGenomes = []\n",
    "        self.highestScore = 0.0\n",
    "        self.fittestGenome = None\n",
    "        \n",
    "        # place genomes into species\n",
    "        for g in self.genomes:\n",
    "            foundSpecies = False\n",
    "            for s in self.species:\n",
    "                if Genome.compatibilityDistance(g, s.mascot, C1, C2, C3) < DT:\n",
    "                    s.members.append(g)\n",
    "                    self.speciesMap[g] = s\n",
    "                    foundSpecies = True\n",
    "                    break\n",
    "            if not foundSpecies:\n",
    "                newSpecies = Species(g)\n",
    "                self.species.append(newSpecies)\n",
    "                self.speciesMap[g] = newSpecies\n",
    "        \n",
    "        # remove unused species\n",
    "        self.species = [s for s in self.species if len(s.members)]\n",
    "        \n",
    "        # evaluate genomes and assign fitness\n",
    "        for g in self.genomes:\n",
    "            s = speciesMap[g]\n",
    "            \n",
    "            score = self.evaluateGenome(g)\n",
    "            adjustedScore = score / len(s.members)\n",
    "            \n",
    "            s.addAdjustedFitness(adjustedFitness)\n",
    "            s.fitnessPop(FitnessGenome(g, adjustedScore))\n",
    "            self.scoreMap[g] = adjustedScore\n",
    "            \n",
    "            if score > self.highestScore:\n",
    "                self.highestScore = score\n",
    "                self.fittestGenome = g\n",
    "        \n",
    "        # put best genomes from each species into next generation\n",
    "        for s in species:\n",
    "            s.fitnessPop.sort(key=operator.attrgetter('fitness'))\n",
    "            fittestInSpecies = s.fitnessPop[-1]\n",
    "            self.nextGenGenomes.append(fittestInSpecies.genome)\n",
    "        \n",
    "        # breed the rest of the genomes\n",
    "        while len(self.nextGenGenomes) < self.populationSize:\n",
    "            s = self.getRandomSpeciesBiasedAdjustedFitness()\n",
    "            child = None\n",
    "            \n",
    "            if random.uniform(0, 1.0) < self.CROSSOVER_RATE:\n",
    "                p1 = self.getRandomGenomeBiasedAdjustedFitness(s)\n",
    "                p2 = self.getRandomGenomeBiasedAdjustedFitness(s)\n",
    "                if self.scoreMap[p2] > self.scoreMap[p1]:\n",
    "                    p1, p2 = p2, p1\n",
    "                child = Genome.crossover(p1, p2)\n",
    "            else:\n",
    "                child = deepcopy(self.getRandomGenomeBiasedAdjustedFitness(s))\n",
    "            \n",
    "            if random.uniform(0, 1.0) < self.MUTATION_RATE:\n",
    "                child.mutation()\n",
    "            if random.uniform(0, 1.0) < self.ADD_CONNECTION_RATE:\n",
    "                child.addConnectionMutation(connectionInnovation, 10)\n",
    "            else if random.uniform(0, 1.0) < self.ADD_NODE_RATE:\n",
    "                child.addNodeMutation(connectionInnovation, nodeInnovation)\n",
    "            \n",
    "            self.nextGenGenomes.append(child)\n",
    "        \n",
    "        self.genomes = self.nextGenGenomes\n",
    "        self.nextGenGenomes = []\n",
    "    \n",
    "    def getRandomSpeciesBiasedAdjustedFitness(self):\n",
    "        completeWeight = sum([s.totalAdjustedFitness for s in self.species])\n",
    "        \n",
    "        r = random.uniform(0, completeWeight)\n",
    "        countWeight = 0.0\n",
    "        for s in self.species:\n",
    "            countWeight += s.totalAdjustedFitness\n",
    "            if countWeight >= r:\n",
    "                return s\n",
    "        \n",
    "        raise Exception(\"Couldn't find a species... \" + len(self.species))\n",
    "       \n",
    "    def getRandomGenomeBiasedAdjustedFitness(self, selectFrom: Species):\n",
    "        completeWeight = sum([fg.fitness for fg in selectFrom.fitnessPop])\n",
    "        \n",
    "        r = random.uniform(0, completeWeight)\n",
    "        countWeight = 0.0\n",
    "        for fg in selectFrom.fitnessPop:\n",
    "            countWeight += fg.fitness\n",
    "            if countWeight >= r:\n",
    "                return fg.genome\n",
    "        \n",
    "        raise Exception(\"Couldn't find a genome... \" + len(selectFrom.fitnessPop))\n",
    "    \n",
    "    @abstractmethod\n",
    "    def evaluateGenome(genome: Genome):\n",
    "        pass\n",
    "    \n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = {1: \"4\", 4: \"3\"}\n",
    "6 in a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abs(-4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
