{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Anna Urbala\n",
    "# Implementacja NEAT\n",
    "\n",
    "Z racji tego, że neat jest skomplikowanym podejściem, jego implementacja rozbita jest na dwa tygodnie. W ramach pierwszego deadline należy przesłać niekoniecznie kompletny i działający kod, który będzie dowodem rozpoczęcia pracy nad rozwiązaniem. Dopiero kod w kolejnym tygodniu musi w pełni działać.\n",
    "\n",
    "W tym tygodniu trzeba zaimplementować: kodowanie, w szczególności generowanie fenotypu z genotypu, wyliczanie funkcji dopasowania i część operatora mutacji polegająca na dodaniu neuronu i zmianie wagi.\n",
    "\n",
    "W ramach testu przeprowadzić próbne uczenie (bez krzyżowania) na zbiorze easy z laboratorium o sieciach neuronowych.\n",
    "\n",
    "https://github.com/gary-butler/NEAT\n",
    "\n",
    "https://towardsdatascience.com/neuro-evolution-on-steroids-82bd14ddc2f6\n",
    "\n",
    "https://github.com/SirBob01/NEAT-Python/blob/master/neat/neat.py\n",
    "\n",
    "https://github.com/goktug97/NEAT/blob/master/neat/neat.py\n",
    "\n",
    "https://github.com/vishnugh/evo-NEAT/tree/master/src/com/evo/NEAT\n",
    "\n",
    "https://www.youtube.com/watch?v=1I1eG-WLLrY\n",
    "\n",
    "http://nn.cs.utexas.edu/downloads/papers/stanley.cec02.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "from enum import Enum, auto\n",
    "import random\n",
    "from copy import deepcopy\n",
    "\n",
    "class NodeType(Enum):\n",
    "    INPUT = auto()\n",
    "    HIDDEN = auto()\n",
    "    OUTPUT = auto()\n",
    "\n",
    "class ConnectionGene:\n",
    "    def __init__(self, inNode: int, outNode: int, weight=1.0, expressed=True, innovation=0):\n",
    "        self.inNode = inNode\n",
    "        self.outNode = outNode\n",
    "        self.weight = weight\n",
    "        self.expressed = expressed\n",
    "        self.innovation = innovation\n",
    "    \n",
    "    def disable(self):\n",
    "        self.expressed = False\n",
    "\n",
    "class NodeGene:\n",
    "    def __init__(self, type_: NodeType, id_: int):\n",
    "        self.type = type_\n",
    "        self.id = id_\n",
    "\n",
    "class Counter:\n",
    "    def __init__(self):\n",
    "        self.innovation = 0\n",
    "        \n",
    "    def getInnovation(self):\n",
    "        self.innovation += 1\n",
    "        return self.innovation\n",
    "    \n",
    "class Genome:\n",
    "    PROB_PERTURBING = 0.9\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.connections = {}\n",
    "        self.nodes = {}\n",
    "    \n",
    "    def addNodeGene(self, gene: NodeGene):\n",
    "        self.nodes[gene.id] = gene\n",
    "    \n",
    "    def addConnectionGene(self, gene: ConnectionGene):\n",
    "        self.connections[gene.innovation] = gene\n",
    "        \n",
    "    def mutation(self):\n",
    "        for conn in self.connections.values():\n",
    "            if random.uniform(0, 1.0) < self.PROB_PERTURBING:\n",
    "                conn.weight *= random.uniform(-2.0, 2.0)\n",
    "            else:\n",
    "                conn.weight = random.uniform(-2.0, 2.0)\n",
    "            \n",
    "    \n",
    "    def addConnectionMutation(self, innovation: Counter):\n",
    "        node1 = self.nodes[random.ranrange(len(self.nodes))]\n",
    "        node2 = self.nodes[random.ranrange(len(self.nodes))]\n",
    "        \n",
    "        rev = False\n",
    "        if node1.type.value > node2.type.value:\n",
    "            rev = True\n",
    "            node1,node2 = node2,node1\n",
    "        \n",
    "        connectionExists = False\n",
    "        for conn in self.connections.values():\n",
    "            if (conn.inNode == node1.id and conn.outNode == node2.id) \\\n",
    "            or (conn.inNode == node2.id and conn.outNode == node1.id):\n",
    "                connectionExists = True\n",
    "                break\n",
    "        \n",
    "        if not connectionExists:\n",
    "            newConn = ConnectionGene(node1.id, node2.id, weight=random.uniform(-1.0, 1.0), innovation=innovation.getInnovation())\n",
    "            self.addConnectionGene(newConn)\n",
    "    \n",
    "    def addNodeMutation(self, innovation: Counter):\n",
    "        conn = self.connections[random.ranrange(len(self.connections))]\n",
    "        \n",
    "        inNode = self.nodes[conn.inNode]\n",
    "        outNode = self.nodes[conn.outNode]\n",
    "        \n",
    "        conn.disable()\n",
    "        \n",
    "        newNode = NodeGene(NodeType.HIDDEN, len(self.nodes))\n",
    "        newConn1 = ConnectionGene(inNode.id, newNode.id, innovation=innovation.getInnovation())\n",
    "        newConn2 = ConnectionGene(newNode.id, outNode.id, weight=conn.weight, innovation=innovation.getInnovation())\n",
    "        \n",
    "        self.addNodeGene(newNode)\n",
    "        self.addConnectionGene(newConn1)\n",
    "        self.addConnectionGene(newConn2)\n",
    "    \n",
    "    @staticmethod\n",
    "    def crossover(parent1: Genome, parent2: Genome):\n",
    "        child = Genome()\n",
    "        for parent1Node in parent1.nodes.values():\n",
    "            child.nodes.append(deepcopy(parent1Node))\n",
    "        \n",
    "        for parent1Conn in paren1.connections.values():\n",
    "            childConnGene = deepcopy(parent1Conn)\n",
    "            if parent1Conn.innovation in parent2.connections and random.choice([True, False]):\n",
    "                childConnGene = deepcopy(parent2.connections[parent1Conn.innovation])\n",
    "            child.addConnectionGene(childConnGene)\n",
    "            \n",
    "        return child\n",
    "    \n",
    "    @staticmethod\n",
    "    def compatibilityDistance(genome1: Genome, genome2: Genome, c1: int, c2: int, c3: int):\n",
    "        excessGenes = Genome.countExcessGenes(genome1, genome2)\n",
    "        disjointGenes = Genome.countDisjointGenes(genome1, genome2)\n",
    "        avgWeightDiff = Genome.averageWeightDiff(genome1, genome2)\n",
    "        return excessGenes * c1 + disjointGenes * c2 + avgWeightDiff * c3\n",
    "    \n",
    "    @staticmethod\n",
    "    def countMatchingGenes(genome1: Genome, genome2: Genome):\n",
    "        matchingGenes = 0\n",
    "        \n",
    "        highestInnovation1 = max(genome1.nodes.keys())\n",
    "        highestInnovation2 = max(genome2.nodes.keys())\n",
    "        indices = max(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices + 1):\n",
    "            if i in genome1.nodes and i in genome2.nodes:\n",
    "                matchingGenes += 1\n",
    "        \n",
    "        highestInnovation1 = max(genome1.connections.keys())\n",
    "        highestInnovation2 = max(genome2.connections.keys())\n",
    "        indices = max(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices + 1):\n",
    "            if i in genome1.connections and i in genome2.connections:\n",
    "                matchingGenes += 1\n",
    "        \n",
    "        return matchingGenes\n",
    "    \n",
    "    @staticmethod\n",
    "    def countDisjointGenes(genome1: Genome, genome2: Genome):\n",
    "        disjointGenes = 0\n",
    "        \n",
    "        highestInnovation1 = max(genome1.nodes.keys())\n",
    "        highestInnovation2 = max(genome2.nodes.keys())\n",
    "        indices = min(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices + 1):\n",
    "            if (i not in genome1.nodes and i in genome2.nodes) \\\n",
    "            or (i not in genome2.nodes and i in genome1.nodes):\n",
    "                disjointGenes += 1\n",
    "        \n",
    "        highestInnovation1 = max(genome1.connections.keys())\n",
    "        highestInnovation2 = max(genome2.connections.keys())\n",
    "        indices = min(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices + 1):\n",
    "            if (i not in genome1.connections and i in genome2.connections) \\\n",
    "            or (i not in genome2.connections and i in genome1.connections):\n",
    "                disjointGenes += 1\n",
    "        \n",
    "        return disjointGenes\n",
    "    \n",
    "    @staticmethod\n",
    "    def countExcessGenes(genome1: Genome, genome2: Genome):\n",
    "        excessGenes = 0\n",
    "        \n",
    "        highestInnovation1 = max(genome1.nodes.keys())\n",
    "        highestInnovation2 = max(genome2.nodes.keys())\n",
    "        indices1 = min(highestInnovation1, highestInnovation2)\n",
    "        indices2 = max(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices1+1, indices2+1):\n",
    "            if (i not in genome1.nodes and i in genome2.nodes) \\\n",
    "            or (i not in genome2.nodes and i in genome1.nodes):\n",
    "                excessGenes += 1\n",
    "        \n",
    "        highestInnovation1 = max(genome1.connections.keys())\n",
    "        highestInnovation2 = max(genome2.connections.keys())\n",
    "        indices1 = min(highestInnovation1, highestInnovation2)\n",
    "        indices2 = max(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices1+1, indices2+1):\n",
    "            if (i not in genome1.connections and i in genome2.connections) \\\n",
    "            or (i not in genome2.connections and i in genome1.connections):\n",
    "                excessGenes += 1\n",
    "        \n",
    "        return excessGenes\n",
    "    \n",
    "    @staticmethod\n",
    "    def averageWeightDiff(genome1: Genome, genome2: Genome):\n",
    "        matchingGenes = 0\n",
    "        weightDifference = 0.0\n",
    "\n",
    "        highestInnovation1 = max(genome1.connections.keys())\n",
    "        highestInnovation2 = max(genome2.connections.keys())\n",
    "        indices = max(highestInnovation1, highestInnovation2)\n",
    "        \n",
    "        for i in range(indices + 1):\n",
    "            if i in genome1.connections and i in genome2.connections:\n",
    "                matchingGenes += 1\n",
    "                weightDifference += abs(genome1.connections[i].weight - genome2.connections[i].weight)\n",
    "        \n",
    "        return weightDifference/matchingGenes\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = {1: \"4\", 4: \"3\"}\n",
    "6 in a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abs(-4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
